File: java/L2000/LeetCode2034.java
package L2000;

import java.util.HashMap;
import java.util.TreeMap;

public class LeetCode2034 {}

class StockPrice {

    private HashMap<Integer, Integer> map;
    private TreeMap<Integer, Integer> cnt;
    private int curTime;

    public StockPrice() {
        map = new HashMap<>();
        cnt = new TreeMap<>();
    }
    
    public void update(int timestamp, int price) {
        if (map.containsKey(timestamp)) {
            Integer p = map.get(timestamp);
            if (cnt.get(p) == 1) {
                cnt.remove(p);
            } else {
                cnt.put(p, cnt.get(p) - 1);
            }
        }
        cnt.put(price, cnt.getOrDefault(price, 0) + 1);
        if (timestamp >= curTime) {
            curTime = timestamp;
        }
        map.put(timestamp, price);
    }
    
    public int current() {
        return map.get(curTime);
    }
    
    public int maximum() {
        return cnt.lastKey();
    }
    
    public int minimum() {
        return cnt.firstKey();
    }
}

File: cpp/L2000/LeetCode2034.cc
#include <map>

class StockPrice {
    int timestamp, price;
    std::map<int, int> timePrices, pricesCount;
public:
    StockPrice(): timestamp(0), price(0) {}
    
    void update(int timestamp, int price) {
        if (timestamp >= this->timestamp) {
            this->timestamp = timestamp;
            this->price = price;
        }
        auto it{timePrices.find(timestamp)};
        if (it != timePrices.end()) {
            pricesCount[it->second] -= 1;
            if (pricesCount[it->second] <= 0) {
                pricesCount.erase(it->second);
            }
            it->second = price;
            pricesCount[price] += 1;
        } else {
            timePrices[timestamp] = price;
            pricesCount[price] += 1;
        }
    }
    
    int current() {
        return this->price;
    }
    
    int maximum() {
        return pricesCount.rbegin()->first;
    }
    
    int minimum() {
        return pricesCount.cbegin()->first;
    }
};